daphne/src/roles/helper.rs

```diff
@@ -216,42 +217,104 @@ pub trait DapHelper<S>: DapAggregator<S> {
         // The reason we do this is because we don't expect replays to happen but we have to guard
         // against them, as such, even though retrying is possibly very expensive, it probably
         // won't happen often enough that it matters.
-        let (out_shares_count, agg_job_resp) = {
-            let mut replayed_reports = HashSet::new();
-            let mut retry_count = 3;
-            loop {
-                if retry_count < 1 {
-                    // we need to prevent an attacker from keeping this loop running for too long,
-                    // potentialy enabling an DOS attack.
-                    return Err(DapAbort::BadRequest(
-                        "AggregationJobContinueReq contained too many replays".into(),
-                    ));
+        const RETRY_COUNT: u32 = 3;
+        let mut input_transitions = agg_job_cont_req.transitions;
+        let mut output_transitions: Vec<(ReportId, Option<TransitionVar>)> = input_transitions
+            .iter()
+            .map(|t| &t.report_id)
+            .cloned()
+            .map(|id| (id, None))
+            .collect::<Vec<_>>();
+        for _ in 0..RETRY_COUNT {
+            let (agg_share_span, agg_job_resp) = task_config.vdaf.handle_agg_job_cont_req(
+                task_id,
+                task_config,
+                &state,
+                agg_job_cont_req.round,
+                &agg_job_id,
+                &input_transitions,
+            )?;
+
+            let put_shares_result = self
+                .try_put_agg_share_span(task_id, task_config, agg_share_span)
+                .await;
+
+            for t in agg_job_resp.transitions {
```

Comment:
You should be able to accomplish this with `iter_mut()` and `zip()` since `aggJ_job_resp` has the same sequence of reports and in the same order. (Otherwise there is a bug.)

Reply:
They have to be in the same order but it's not the same sequence. So I can't use zip but I can make this O(n) instead of O(n^2)

Reply:
How might the sequences differ? I would expect `agg_job_resp.transitions[i].report_id` should be equal to `output-transitions[i].report_id` for all `i`.